File: engine_core/src/maps/map_2d_loader.rs
use std::collections::HashMap;
use std::fs;

use crate::maps::{map_2d_model::Map2DModel, map_json::MapJson};

pub fn load_map_from_json(path: &str) -> Map2DModel {
    let content = fs::read_to_string(path).expect("Failed to read map file");
    let map_json: MapJson =
        serde_json::from_str(&content).expect("Failed to parse JSON map file");

    let data: Vec<String> = map_json
        .map_data
        .lines()
        .map(|line| line.to_string())
        .collect();

    let symbols: HashMap<String, String> = map_json.symbols;
    let textures_path = map_json.textures_path;
    let textures = map_json.textures;

    Map2DModel {
        tile_size: map_json.tile_size,
        data,
        symbols,
        textures_path,
        textures,
    }
}

File: GameEngineLab.Core/Features/Maps/Systems/Map2DLoader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using GameEngineLab.Core.Features.Maps.Resources;

namespace GameEngineLab.Core.Features.Maps.Systems;

public static class Map2DLoader
{
    public static Map2DModel LoadFromJson(string path)
    {
        if (string.IsNullOrWhiteSpace(path))
        {
            throw new ArgumentException("Map path is required.", nameof(path));
        }

        if (!File.Exists(path))
        {
            throw new FileNotFoundException("Map file was not found.", path);
        }

        var json = File.ReadAllText(path);
        var model = JsonSerializer.Deserialize<MapJsonDto>(json, JsonOptions)
            ?? throw new InvalidDataException("Map JSON could not be parsed.");

        var data = model.MapData
            .Split('\n', StringSplitOptions.None)
            .Select(static line => line.TrimEnd('\r'))
            .ToArray();

        var textures = model.Textures.ToDictionary(
            static pair => pair.Key,
            static pair => MapTextureSource.FromRaw(pair.Value),
            StringComparer.OrdinalIgnoreCase);

        return new Map2DModel
        {
            TileSize = model.TileSize,
            Data = data,
            Symbols = new Dictionary<string, string>(model.Symbols, StringComparer.OrdinalIgnoreCase),
            TexturesPath = model.TexturesPath,
            Textures = textures,
        };
    }

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        ReadCommentHandling = JsonCommentHandling.Skip,
        AllowTrailingCommas = true,
    };

    private sealed class MapJsonDto
    {
        [JsonPropertyName("tile_size")]
        public byte TileSize { get; init; }

        [JsonPropertyName("symbols")]
        public Dictionary<string, string> Symbols { get; init; } = new(StringComparer.OrdinalIgnoreCase);

        [JsonPropertyName("textures_path")]
        public string TexturesPath { get; init; } = string.Empty;

        [JsonPropertyName("textures")]
        public Dictionary<string, string> Textures { get; init; } = new(StringComparer.OrdinalIgnoreCase);

        [JsonPropertyName("map_data")]
        public string MapData { get; init; } = string.Empty;
    }
}
